metals/src/main/scala/scala/meta/internal/metals/FormattingProvider.scala

```diff
@@ -0,0 +1,93 @@
+package scala.meta.internal.metals
+
+import scala.meta._
+import org.eclipse.{lsp4j => l}
+import MetalsEnrichments._
+import scala.language.reflectiveCalls
+import java.util
+import java.nio.file.Files
+import scala.util.Try
+
+/**
+ * Implement text formatting using Scalafmt
+ */
+final class FormattingProvider(
+    workspace: AbsolutePath,
+    buffers: Buffers,
+    embedded: Embedded,
+    userConfig: () => UserConfiguration
+) {
+
+  def format(path: AbsolutePath): util.List[l.TextEdit] = {
+    val input = path.toInputFromBuffers(buffers)
+    val fullDocumentRange = Position.Range(input, 0, input.chars.length).toLSP
+    (for {
+      scalafmt <- classloadedScalafmt
+      formatted <- scalafmtConfigPath match {
+        case Some(configPath) =>
+          scalafmt
+            .format(input.text, configPath.toString, input.path)
+            .fold(e => {
+              scribe.error("Error while running Scalafmt", e)
+              None
+            }, Some(_))
+        case None if !userConfig().scalafmtRequireConfigFile =>
+          scalafmt
+            .format(input.text, input.path)
+            .fold(e => {
+              scribe.error("Error while running Scalafmt", e)
+              None
+            }, Some(_))
+        case _ =>
+          scribe.info(
+            "Skipping formatting request because no Scalafmt config file is present"
+          )
+          None
+      }
+    } yield {
+      List(new l.TextEdit(fullDocumentRange, formatted))
+    }).getOrElse(Nil).asJava
+  }
+
+  private def scalafmtConfigPath: Option[AbsolutePath] = {
+    val relativePath =
+      userConfig().scalafmtConfigPath.getOrElse(".scalafmt.conf")
+    Some(workspace.resolve(relativePath))
+      .filter(path => Files.isRegularFile(path.toNIO))
+  }
+
+  private val classloadedScalafmt: Option[Scalafmt] = {
```

Comment:
I know nothing about "classloading best practices", so I just blindly ported the implementation we had for 0.1.0.

There may be better ways to do this.

Reply:
This should not be a `val` since the scalafmt version may change. I would make it a `Collections.synchronizedMap[String, Scalafmt]` and fetch on demand. This means the first formatting request may take a while, but we don't slow down server startup.

Reply:
I thought about this, and I assumed it would be ok to restart the server if the Scalafmt version changes.

This is actually a nicer idea. I'll work on it 👍

Reply:
Most critically, we should not slow down server startup by downloading Scalafmt. Only classload it if used.